Ss4 improvements

bummzack opened the pull request:
Here's the current state of the changes.

- Moved from two individual fields to a composite-field for the focus.
- Added migration-script to up- and downgrade.
- Moved files into matching folders, updated namespaces.

So far, everything seems to work as it did before and tests pass.
Feel free to review this changeset and merge if you feel comfortable with it.

I would hold on to releasing this though. I think it would be worthwhile to also transition to a react component and I'll start working on that. Some questions for you:

- How do you feel about maintaining the react component in the upcoming future? 
- Can we transition from CSS to SCSS?
- What about building the crosshair and gradient stuff entirely in CSS? Seems to be supported in pretty much every recent browser (https://caniuse.com/#search=radial-gradient) and SS4 CMS has a requirement of IE11+
- Maybe we could also double the resolution of the preview-image (for Hi-DPI displays)

jonom commented:
>How do you feel about maintaining the react component in the upcoming future?

Some of the intricacies of React confuse me but I'm all good with the basics, so should be okay.

> Can we transition from CSS to SCSS?

If there's enough CSS to justify it. I love SCSS but I don't like to introduce build tools if they're not critical as makes it harder for contributors to get involved. For the amount of css in the project currently I don't think it would be worth introducing a pre-processor.

> What about building the crosshair and gradient stuff entirely in CSS?

Hmm i don't see why not! I would probably want the crosshair to be constructed with css rather than a **+** character though just to ensure it's pixel perfect and as close to dead centre as possible. Probably not worth the trouble to be honest, but you could base64 encode an image in the CSS for that part.

> Maybe we could also double the resolution of the preview-image (for Hi-DPI displays)

Go for it

bummzack commented:
Almost there… I just have a litte trouble wrapping my head around the react-forms stuff :)

Regarding CSS: There's a whole toolchain involved with react anyway and it's set up to also compile SCSS. Sure, the build tools are more complicated, but it would be in-line with how silverstripe core modules do it.

I use CSS gradients and an SVG sprite for the crosshair currently. So far it looks good.

jonom commented on src/Extensions/FocusPointImageExtension.php:
```diff
@@ -0,0 +1,128 @@
+     */
+    public function PercentageX()
+    {
+        return round($this->focusCoordToOffset('x', $this->owner->getField('FocusPoint')->FocusX) * 100);
```
You can do something like `$Image.Fit(200,200).DebugFocusPoint` to trigger this error

jonom commented on src/FieldType/DBFocusPoint.php:
```diff
@@ -0,0 +1,268 @@
+        if ($image->isSize($width, $height) && !Config::inst()->get(DBFile::class, 'force_resample')) {
+            return $image;
+        } elseif ($cropData = $this->calculateCrop($width, $height, $imgW, $imgH)) {
+            $variant = $image->variantName(__FUNCTION__, $width, $height, $cropData['CropAxis'], $cropData['CropOffset']);
```
This is a picky thing but now that this code is in a function called generateFocusFill, `__FUNCTION__` is returning that and so our resampled images are prefixed with 'generateFocusFill'. Can you replace `__FUNCTION__` here with 'FocusFill' please.

bummzack replied:
Good point

jonom commented on src/FieldType/DBFocusPoint.php:
```diff
@@ -0,0 +1,268 @@
+                    return null;
+                }
+
+                // Update FocusPoint
```
This part might need tweaking to support the new field spec

jonom commented:
@bummzack if you find yourself looking for even more to do (😄), this one is a release blocker:

> When working with images in the CMS outside of the Files section (e.g. HTMLEditorField insert), the editing form is disabled for some reason. I assume maybe this will be restored at some point? Currently the FocusPoint field is still active in that context, but you can't save the changes. This is a very bad experience for the user so FocusPoint field probably needs a read only or disabled option added.

bummzack commented on src/Forms/FocusPointField.php:
```diff
@@ -42,21 +40,29 @@ class FocusPointField extends FieldGroup
+            $previewImage = $image->FitMax($this->config()->get('max_width'), $this->config()->get('max_height'));
+            array_unshift($fields, LiteralField::create('FocusPointGrid', $previewImage->renderWith(
+                FocusPointField::class,
+                ['FieldGridBackgroundCSS' => $image->FieldGridBackgroundCSS($previewImage->getWidth(), $previewImage->getHeight())]
```
Somehow the template was rendered with a `DBFile` context and didn't have the focuspoint set… I imagine the focuspoint was lost during `FitMax`.
Not sure what's a good way to circumvent that. Overall, the template rendering in this instance is moot, as it'll be superseded by the react component.

bummzack replied:
Ah, I know what the issue was. Somehow the `FocusPointImageExtension` was also applied to `DBFile`. That's not really a good idea, as `DBFile` should already have access to the image props/methods via failover. I'll investigate further.

jonom replied:
`DBFile` uses the `ImageManipulation` trait, so presumably it needs to receive most (not all) of the properties and methods in `FocusPointImageExtension`. I'm guessing that resampled image objects are `DBFiles` rather than `Images` in some contexts but I got lost trying to understand that.

bummzack replied:
`ImageManipulation` always sets the original file on the `DBFile` after a manipulation. So in `ImageManipulation::manipulate`, when it created a new variant of an `Image`, it'll return a `DBFile` instance with the original Image as failover.

So any method you call on `DBFile` that doesn't exist there will call the method on the original. Therefore an additional field or method should be inherited automatically.

jonom replied:
> So any method you call on DBFile that doesn't exist there will call the method on the original. Therefore any additional field or method should be inherited automatically.

Interesting. Bear with me, this concept is new to me. So it would call the method on the original Image object? That would interfere with chaining then right? Like you couldn't go `$Image.Greyscale.FocusFill(50,50)` for example because you would just end up with `$Image.FocusFill(50,50)`?

I also want to pass through updated FocusPoint coords after a crop, adjusted to represent the same point in the cropped image. This is important for me in case you want to do anything with the focus point in the front end, for instance set the css transform point to the same location, or integrate with scripts like jQuery FocusPoint.

bummzack commented on src/InterventionBackend.php:
```diff
@@ -1,54 +0,0 @@
-<?php
```
Invalidating cache with params to `variantName` seems to work fine already. I'm not too sure if overriding the default image-manipulation methods is actually a good idea though? What if I want the original way to crop an image? I think having specific method names is better to be honest…

We could also integrate some of the methods on the `DBFocusPoint` field. So it could  have a `Fill` method and you would then invoke it via `$Image.FocusPoint.Fill`… still more to write than `$Image.FocusFill` though ;)

jonom replied:
> Invalidating cache with params to variantName seems to work fine already.

Yep it works fine for FocusFill, but not if you're replacing the built in methods. Replacing the built in methods is something I'm very keen on, but it could be optional. I would love FocusPoint to be a drop in replacement for cropping and can't think of any instance where a centred crop would be better than a focused one. This would mean things like CMS thumbnails benefit from focus based cropping and any themes or templates that use built in cropping methods wouldn't need to be overridden to get upgraded cropping.

jonom replied:
I haven't tested yet but I understand a trait works as if the code in it has been copy-pasted in to whatever is using it. So if I put an extension point in that method I should be able to apply an extension to Image and DBFile (the two classes that use it) to hook in to it.

bummzack commented on src/FieldType/DBFocusPoint.php:
```diff
@@ -0,0 +1,268 @@
+     * FocusY: Decimal number between -1 & 1, where -1 is bottom, 0 is center, 1 is top.
+     */
+    private static $composite_db = [
+        'FocusX' => 'Double',
```
yeah, why not.

bummzack commented:
I worked some more on this and updated this PR. All the issues should be resolved now and I also added some more test-cases.
I switched the coordinate-system, since it seemed to be a lot clearer to do this:

	-1,-1 -- 0 -- 1,-1
	  |            |
	  |            |
	  0     0,0    0
	  |            |
	  |            |
	-1,1 --- 0 -- 1,1

But I notice that this would break compatibility with your jQuery module, so I'm going to revert that.

bummzack commented:
Flipped the coordinate system back to:

	-1,1 --- 0 -- 1,1
	  |            |
	  |            |
	  0     0,0    0
	  |            |
	  |            |
	-1,-1 -- 0 -- 1,-1

jonom commented:
This is a heck of a pull request 😀  

I'm not too worried about the coordinate system. In the latest version of that jQuery plugin (which will probably never see the light of day) I switched the coord system to match CSS e.g. 0,0 top left to 100,100 bottom right because you're right, it's just easier. (When I first devised this concept I just used a standard plotting grid). I want to keep 0,0 in the centre for this module because it's a reliable default (null would resolve to centre) but I don't mind if you flip the vertical axis.

Can you add some documentation about the build  process and any steps that need to be followed for contributing and publishing changes?

I'll try to test this thoroughly in the next few days.

Please feel free to add your details in the author section of the composer.json file.

bummzack commented:
Thanks for the feedback. I can change it back to the other coordinate system, no problem :) I think having the center denoted by `0,0` is a good idea in this case, since the default value of a `double` or `float` field in silverstripe is `0` (thus the focus-point will be centered by default).

I didn't research further into overriding the core manipulation methods. But you hit a wall already with these and I guess we can bring back your custom backend when the issue gets resolved.

Will add some documentation about building soon. Here's the gist of it:
```bash
cd ./vendor/jonom/focuspoint
yarn # to install the npm dependencies
yarn run dev # to build a dev-release of js&css
yarn run build # to build production releases of js&css (eg. before you commit)
yarn run watch # automatically rebuild whenever you change a file. Useful for development
```

bummzack commented:
Flipped back the coordinate system and added documentation about the build-tools.

jonom commented:
Looks good to me @bummzack, thanks for this huge effort! I've tagged a 3.0.0-beta.1 release for now, mostly because I wanted to try tagging a pre-release. It seems stable to me so unless I hear any complaints I'll tag 3.0 in a week or so.

bummzack commented:
Awesome! Let's hope for many testers (I'm already running it in a website that went live today and didn't get any complaints).
